src/humancellatlas/data/metadata/lookup.py

```diff
@@ -18,48 +110,115 @@ def lookup(d: Mapping[K, V], k: K, *ks: K, default: Union[V, LookupDefault] = Lo
     that argument instead of raising an KeyError in that case. This is notably different to dict.get() whose default
     default is `None`. This function does not have a default default.
 
+    :param d: a dict
+    :param k: a fully qualified key to lookup in the dict eg. 'publications.title'
+    :param ks: optional additional fully qualified key(s) to try if the first key is not found
+    :param default: a value to return if no key is found in the dict
+    :return: the value that was found in dict d, or if not found either the default is returned or a KeyError is raised
+
     If the first key is present, return its value ...
-    >>> lookup({1:2}, 1)
-    2
+    >>> lookup({'a': 'b'}, 'schema.a')
+    'b'
 
     ... and ignore the other keys.
-    >>> lookup({1:2}, 1, 3)
-    2
+    >>> lookup({'a': 'b'}, 'schema.a', 'schema.X')
+    'b'
 
     If the first key is absent, try the fallbacks.
-    >>> lookup({1:2}, 3, 1)
-    2
+    >>> lookup({'a': 'b'}, 'schema.X', 'schema.a')
+    'b'
 
     If the key isn't present, raise a KeyError referring to that key.
-    >>> lookup({1:2}, 3)
+    >>> lookup({'a': 'b'}, 'schema.X')
     Traceback (most recent call last):
     ...
-    KeyError: 3
+    KeyError: 'X'
 
     If neither the first key nor the fallbacks are present, raise a KeyError referring to the first key.
-    >>> lookup({1:2}, 3, 4)
+    >>> lookup({'a': 'b'}, 'schema.X', 'schema.Y')
     Traceback (most recent call last):
     ...
-    KeyError: 3
+    KeyError: 'X'
 
     If the key isn't present but a default was passed, return the default.
-    >>> lookup({1:2}, 3, default=4)
-    4
+    >>> lookup({'a': 'b'}, 'schema.X', default='c')
+    'c'
 
     None is a valid default.
-    >>> lookup({1:2}, 3, 4, default=None) is None
+    >>> lookup({'a': 'b'}, 'schema.X', default=None) is None
     True
     """
+
+    if 'schema_version' in d or 'describedBy' in d:
+        version = get_document_version(d)
+        new_k = schema_template.replaced_by_at(k, version)
+    else:
+        version = '0.0.0'
+        new_k = k
+
+    # HACK for missing entries in https://schema.dev.data.humancellatlas.org/property_migrations
```

Comment:
@simonjupp, do you happen to know if these migrations were added?

Reply:
No, I think we only have migrations from when the migration file was introduced. Do you think it’s worth retrofitting now? If these are really old I’d say no, if they are needed for data currently in production then we can probably put them in.